### ISSUE TITLE
different eval/loss when training on multiple GPUs vs single GPU

### ISSUE THREAD

**Issue by geronimi73** [huggingface/accelerate#2393]
I have an issue when training on multiple GPUs in DataParallel mode. From what I understand, training on 4 GPUs quadruples the effective batch size. Training on a single 1 GPU with batch size 4 should therefore be the same as training on 4 GPUs with BS 1. Right?

I guess I am wrong because the eval/loss is completely different in these two cases. With 1 GPU, the loss curve looks like overfitting, with 4 GPUs eval/loss nicely goes down together with train/loss. 

Is this real or just because the validations loss are calculated on each GPU and averaged or something?

key settings:
* batch size of 1 (4 GPUs) versus batch size of 4 (1 GPU)
* LR the same in both cases, gradient acc. steps always 1
* different seed for each GPU
* phi-2, DataParallel

![W B Chart 30_01_2024, 16_17_28](https://github.com/huggingface/accelerate/assets/141400217/fa5b0903-2402-4700-831a-83f112eb6092)

Code (cut):
```python
accelerator = Accelerator()
set_seed(42 + accelerator.process_index)

modelpath="microsoft/phi-2"
dataset_name="g-ronimo/riddles_evolved"
lr=0.00002
bs=1            # or 4
bs_eval=16       
epochs=20

model = AutoModelForCausalLM.from_pretrained(
    modelpath,    
    device_map={"": accelerator.process_index},
    quantization_config=BitsAndBytesConfig(
        load_in_4bit=True,
        bnb_4bit_compute_dtype=torch.bfloat16,
        bnb_4bit_quant_type="nf4",
    ),
    torch_dtype=torch.bfloat16,
)

tokenizer = AutoTokenizer.from_pretrained(modelpath, use_fast=False)   

tokenizer.add_tokens(["<|im_start|>", "<PAD>"])
tokenizer.pad_token = "<PAD>"
tokenizer.add_special_tokens(dict(eos_token="<|im_end|>"))
model.config.eos_token_id = tokenizer.eos_token_id

model = prepare_model_for_kbit_training(model, use_gradient_checkpointing=False)    # use_gradient_checkpointing not yet working

lora_config = LoraConfig(
    r=32, 
    lora_alpha=32, 
    target_modules = [ "q_proj", "k_proj", "v_proj", "dense" ],
    modules_to_save = ["lm_head", "embed_tokens"],
    lora_dropout=0.1, 
    bias="none", 
    task_type="CAUSAL_LM",
)
model = get_peft_model(model, lora_config)
model.config.use_cache = False

dataset = load_dataset(dataset_name)["train"].train_test_split(test_size=0.1)

def tokenize(input, max_length):
# .. nothing exciting happening here

dataset_tokenized = dataset.map(
    partial(tokenize, max_length=256), 
    batched=False, 
    num_proc=os.cpu_count()//accelerator.num_processes, 
    remove_columns=dataset["train"].column_names
)

def collate(elements):
# .. nothing exciting happening here neither
 
steps_per_epoch=len(dataset_tokenized["train"])//(accelerator.num_processes*bs)

args = TrainingArguments(
    output_dir="out",
    per_device_train_batch_size=bs,
    per_device_eval_batch_size=bs_eval,
    evaluation_strategy="steps",
    logging_steps=1,
    eval_steps=steps_per_epoch//2,    # 2 evals per epoch
    save_steps=steps_per_epoch,     
    gradient_accumulation_steps=1,
    num_train_epochs=epochs,
    lr_scheduler_type="constant",
    optim="paged_adamw_32bit",      
    learning_rate=lr,
    group_by_length=False,
    bf16=True,        
    ddp_find_unused_parameters=False,
)

trainer = Trainer(
    model=model,
    tokenizer=tokenizer,
    args=args,
    data_collator=collate,
    train_dataset=dataset_tokenized["train"],
    eval_dataset=dataset_tokenized["test"],
)

trainer.train()
```

```
accelerate==0.26.1
peft==0.7.1
torch==2.1.0
transformers @ git+https://github.com/huggingface/transformers@de13a951b38b85195984164819f1ab05fe508677
```

thank you for taking a look at this! Would love to know where my mistake is

@muellerzr

**Comment by muellerzr** [comment]
First thing to look at is:

> LR the same in both cases, gradient acc. steps always 1

We step the LR scheduler * n_gpus. So your lr scheduler needs to be adjusted for this: https://huggingface.co/docs/accelerate/concept_guides/performance#learning-rates

Second:

Set a standard seed, don't set it to be individual on each node with `set_seed`.

**Comment by geronimi73** [comment]
thanks for the quick reply! 

> We step the LR scheduler * n_gpus. 

Does this mean, when I pass my LR to trainer, accelerate will automatically multiply this number with the number of GPUs? 

> Set a standard seed, don't set it to be individual on each node with `set_seed`.

If I do that in the code shown then `train_test_split` will lead to exactly the same train/val-split on each GPU and they will all train on exactly the same data in parallel. Which is not what I want, I guess.

**Comment by muellerzr** [comment]
No. You need to handle this yourself.

**Comment by geronimi73** [comment]
OK. But the effective batch size is the same in both cases, single and multiple GPU. Why does the LR have to be adjusted?

**Comment by muellerzr** [comment]
Because the LR scheduler will step based on a scaling of the num processes. So it has to be adjusted (as that documentation states)

**Comment by geronimi73** [comment]
What I didn't mention explicitly (it's just in the code): it's a constant LR `lr_scheduler_type="constant"`

**Comment by geronimi73** [comment]
what I mean is that the LR scheduler step does not matter with a constant scheduler. No?

**Comment by muellerzr** [comment]
Yes indeed. Two things to try:

1. Run `set_seed` before `Accelerator()`
2. Use `use_seedable_sampler=False` in the `Accelerator` (or `True`, whatever the default *isn't*)

**Comment by geronimi73** [comment]
OK I think I got it now. As you pointed out, I was doing something completely stupid.

> Set a standard seed, don't set it to be individual on each node with `set_seed`.

What I think happened was that each process (GPU) generated it's own train/val split and passed it onto the trainer. I simply don't know enough of what's happening in Trainer under the hood, but I assume that each GPU was validating on samples that another GPU was training on, therefore the loss was constantly decreasing because I was basically training on the validation set 🫣

Thank you for your time and attention @muellerzr

**Comment by geronimi73** [comment]
closing

**Comment by geronimi73** [comment]
One last thought. whenever someone calls `train_test_split` on a dataset **without** setting a seed, then they will have the same problem because each process will randomize train/val differently. 

I am not aware of warnings in the docs that you must always set a (common) seed when running multiple processes with `accelerate launch`. There must be a lot of people making this mistake when training on multiple GPUs.
